substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+							let _ = T::TargetList::on_update(who, Self::to_currency(balance))
+								.defensive_proof(
+									"staker exists in the list as per the check above; qed.",
+								);
+							false
+						}
+					} else {
+						defensive!("unexpected: unable to fetch score from staking interface of an existent staker");
+						false
+					}
+				},
+			};
+
+			log!(
+				debug,
+				"update_score of {:?} by {:?}. removed target node? {}",
+				who,
+				imbalance,
+				removed
+			);
+		}
+	}
+}
+
+#[cfg(any(test, feature = "try-runtime"))]
+impl<T: Config> Pallet<T> {
+	/// Try-state checks for the stake-tracker pallet.
+	///
+	/// 1. `do_try_state_approvals`: checks the curent approval stake in the target list compared
+	///    with the staking state.
+	/// 2. `do_try_state_target_sorting`: checks if the target list is sorted by score (approvals).
+	/// 3. `do_try_state_voter_sorting`: checks if the voter list is sorted by score (stake).
+	pub fn do_try_state() -> Result<(), sp_runtime::TryRuntimeError> {
+		Self::do_try_state_approvals()?;
+		Self::do_try_state_target_sorting()?;
+		Self::do_try_state_voter_sorting()?;
+
+		Ok(())
+	}
+
+	/// Try-state: checks if the approvals stake of the targets in the target list are correct.
+	///
+	/// These try-state checks generate a map with approval stake of all the targets based on
+	/// the staking state of stakers in the voter and target lists. In doing so, we are able to
+	/// verify that the current voter and target lists and scores are in sync with the staking
+	/// data and perform other sanity checks as the approvals map is calculated.
+	///
+	/// NOTE: this is an expensive state check since it iterates over all the nodes in the
+	/// target and voter list providers.
+	///
+	/// Invariants:
+	///
+	/// * Target List:
+	///   * The sum of the calculated approvals stake is the same as the current approvals in
+	///   the target list per target.
+	///   * The target score of an active validator is the sum of all of its nominators' stake
+	///   and the self-stake;
+	///   * The target score of an idle validator (i.e. chilled) is the sum of its nominator's
+	///   stake. An idle target may not be part of the target list, if it has no nominations.
+	///   * The target score of a "dangling" target (ie. idle AND unbonded validator) must
+	///   always be > 0. We expect the stake-tracker to have cleaned up dangling targets with 0
+	///   score.
+	///   * The number of target nodes in the target list matches the number of
+	///   (active_validators + idle_validators + dangling_targets_score_with_score).
+	///
+	/// * Voter List:
+	///  * The voter score is the same as the active stake of the corresponding stash.
+	///  * An active validator should also be part of the voter list.
+	///  * An idle validator should not be part of the voter list.
+	///  * A dangling target shoud not be part of the voter list.
+	pub(crate) fn do_try_state_approvals() -> Result<(), sp_runtime::TryRuntimeError> {
```

Comment:
Conceptually, what this code is the same as what the staking pallet is doing upon the MBM migration of all approvals. If that block of code was accessible, I would say we should reuse it here. So `StakingInterface` as `#[cfg(feature = "try-runtime")] fn force_build_approvals() -> Approvals` which creates all approvals, and in here we just compare it with `TargetList`, which should always match.

Reply:
Probably not practical as the migration is MBM.